File: ios/Packages/GemstoneServices/Sources/Stores/ConnectionStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import protocol Gemstone.GemConnectionStore
import struct Gemstone.WalletConnection
import struct Gemstone.WalletConnectionSession
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneConnectionStore: GemConnectionStore, @unchecked Sendable {
    private let store: ConnectionStore

    public init(store: ConnectionStore) {
        self.store = store
    }

    public func getConnection(sessionId: String) async throws -> Gemstone.WalletConnection? {
        try store.getConnection(sessionId: sessionId).map { $0.toGem() }
    }

    public func getSessions() async throws -> [Gemstone.WalletConnectionSession] {
        try store.getSessions().map { $0.toGem() }
    }

    public func addConnection(connection: Gemstone.WalletConnection) async throws {
        try store.addConnection(connection.toPrimitives())
    }

    public func updateSession(session: Gemstone.WalletConnectionSession) async throws {
        try store.updateConnectionSession(session.toPrimitives())
    }

    public func deleteSessions(sessionIds: [String]) async throws {
        _ = try store.delete(ids: sessionIds)
    }
}

File: android/data/services/gemstone/src/main/kotlin/com/gemwallet/android/data/services/gemstone/stores/ConnectionStore.kt
package com.gemwallet.android.data.services.gemstone.stores

import com.gemwallet.android.data.services.store.database.ConnectionsDao
import com.gemwallet.android.data.services.store.database.entities.toDTO
import com.gemwallet.android.data.services.store.database.entities.toRecord
import com.gemwallet.android.data.services.store.database.entities.toSession
import com.gemwallet.android.ext.toGem
import com.gemwallet.android.ext.toPrimitives
import com.wallet.core.primitives.WalletConnection
import kotlinx.coroutines.flow.firstOrNull
import uniffi.gemstone.GemConnectionStore

class GemstoneConnectionStore(private val walletStore: GemstoneWalletStore, private val connectionsDao: ConnectionsDao) : GemConnectionStore {

    suspend fun getConnectionBySessionId(sessionId: String): WalletConnection? {
        val record = connectionsDao.getBySessionId(sessionId) ?: return null
        return record.toDTO(walletStore.observeWallets().firstOrNull().orEmpty())
    }

    override suspend fun getConnection(sessionId: String): uniffi.gemstone.WalletConnection? = getConnectionBySessionId(sessionId)?.toGem()

    override suspend fun getSessions(): List<uniffi.gemstone.WalletConnectionSession> = connectionsDao.getConnections().map { it.toSession().toGem() }

    override suspend fun addConnection(connection: uniffi.gemstone.WalletConnection) = connectionsDao.insert(connection.toPrimitives().toRecord())

    override suspend fun updateSession(session: uniffi.gemstone.WalletConnectionSession) {
        val updated = session.toPrimitives()
        val record = connectionsDao.getBySessionId(updated.id) ?: return
        connectionsDao.insert(
            record.copy(
                state = updated.state,
                chains = updated.chains,
                expireAt = updated.expireAt,
                appName = updated.metadata.name,
                appDescription = updated.metadata.description,
                appUrl = updated.metadata.url,
                appIcon = updated.metadata.icon,
            ),
        )
    }

    override suspend fun deleteSessions(sessionIds: List<String>) = connectionsDao.delete(sessionIds)
}
